cmd/kubeadm/app/preflight/checks.go

```diff
@@ -761,38 +796,49 @@ func RunInitMasterChecks(cfg *kubeadmapi.MasterConfiguration) error {
 }
 
 // RunJoinNodeChecks executes all individual, applicable to node checks.
-func RunJoinNodeChecks(cfg *kubeadmapi.NodeConfiguration) error {
+func RunJoinNodeChecks(execer utilsexec.Interface, cfg *kubeadmapi.NodeConfiguration, criSocket string) error {
 	// First, check if we're root separately from the other preflight checks and fail fast
 	if err := RunRootCheckOnly(); err != nil {
 		return err
 	}
 
+	// check if we can use crictl to perform checks via the CRI
+	criCtlChecker := InPathCheck{executable: "crictl", mandatory: false, exec: execer}
+	warns, _ := criCtlChecker.Check()
+	useCRI := len(warns) == 0
+
 	checks := []Checker{
-		SystemVerificationCheck{},
+		SystemVerificationCheck{CRISocket: criSocket},
 		IsPrivilegedUserCheck{},
 		HostnameCheck{cfg.NodeName},
 		KubeletVersionCheck{},
 		ServiceCheck{Service: "kubelet", CheckIfActive: false},
-		ServiceCheck{Service: "docker", CheckIfActive: true},
 		PortOpenCheck{port: 10250},
 		DirAvailableCheck{Path: filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.ManifestsSubDirName)},
 		FileAvailableCheck{Path: cfg.CACertPath},
 		FileAvailableCheck{Path: filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.KubeletKubeConfigFileName)},
 	}
+	if useCRI {
+		checks = append(checks, CRICheck{socket: criSocket, exec: execer})
+	} else {
+		// assume docker
+		checks = append(checks, ServiceCheck{Service: "docker", CheckIfActive: true})
+	}
 	//non-windows checks
 	if runtime.GOOS == "linux" {
 		checks = append(checks,
 			FileContentCheck{Path: bridgenf, Content: []byte{'1'}},
 			SwapCheck{},
-			InPathCheck{executable: "ip", mandatory: true},
-			InPathCheck{executable: "iptables", mandatory: true},
-			InPathCheck{executable: "mount", mandatory: true},
-			InPathCheck{executable: "nsenter", mandatory: true},
-			InPathCheck{executable: "ebtables", mandatory: false},
-			InPathCheck{executable: "ethtool", mandatory: false},
-			InPathCheck{executable: "socat", mandatory: false},
-			InPathCheck{executable: "tc", mandatory: false},
-			InPathCheck{executable: "touch", mandatory: false})
+			InPathCheck{executable: "ip", mandatory: true, exec: execer},
+			InPathCheck{executable: "iptables", mandatory: true, exec: execer},
+			InPathCheck{executable: "mount", mandatory: true, exec: execer},
+			InPathCheck{executable: "nsenter", mandatory: true, exec: execer},
+			InPathCheck{executable: "ebtables", mandatory: false, exec: execer},
+			InPathCheck{executable: "ethtool", mandatory: false, exec: execer},
+			InPathCheck{executable: "socat", mandatory: false, exec: execer},
+			InPathCheck{executable: "tc", mandatory: false, exec: execer},
+			InPathCheck{executable: "touch", mandatory: false, exec: execer}),
+			criCtlChecker
```

Comment:
I _think_ this results in an compilation error.
Shouldn't CRI be used in Windows though?

Reply:
omg right